zero/src/prover.rs

```diff
@@ -131,29 +135,180 @@ impl BlockProverInput {
+                                    ))?;
+                            };
+                            Ok::<(), anyhow::Error>(())
+                        });
+
+                        segment_proving_tasks.push(segment_proving_task);
+                        segment_counter += 1;
+                    }
+                    drop(segment_proof_tx);
+                    // Wait for all the segment proving tasks of one batch to finish.
+                    while let Some((segment_idx, segment_aggregatable_proof)) = segment_proof_rx.recv().await {
+                        batch_segment_aggregatable_proofs.push((segment_idx, segment_aggregatable_proof));
+                    }
+                    try_join_all(segment_proving_tasks).await?;
+                    batch_segment_aggregatable_proofs.sort_by(|(a, _), (b, _)| a.cmp(b));
+                    debug!(block_number=%block_number, batch=%batch_idx, "finished proving all segments");
+                    // We have proved all the segments in a batch,
+                    // now we need to aggregate them to the batch proof.
+                    // Fold the segment aggregated proof stream into a single batch proof.
+                    let batch_proof = if batch_segment_aggregatable_proofs.len() == 1 {
+                        // If there is only one segment aggregated proof, just transform it to batch proof.
+                        (batch_idx, crate::proof_types::BatchAggregatableProof::from(
+                            batch_segment_aggregatable_proofs.pop().map(|(_, it)| it).unwrap(),
+                        ))
+                    } else {
+                        Directive::fold(IndexedStream::from(batch_segment_aggregatable_proofs.into_iter().map(|(_, it)| it)), &seg_agg_ops)
+                            .run(&proof_runtime.light_proof)
+                            .map(move |e| {
+                                e.map(|p| {
+                                    (
+                                        batch_idx,
+                                        crate::proof_types::BatchAggregatableProof::from(p),
+                                    )
+                                })
+                            })
+                            .await?
+                    };
+                    debug!(block_number=%block_number, batch=%batch_idx, "generated batch proof for block");
+                    batch_proof_tx.send(batch_proof).await.context(format!(
+                        "unable to send batch proof, block: {block_number}, batch: {batch_idx}"
+                    ))?;
+                    Ok::<(), anyhow::Error>(())
+                });
+            }
+            // Wait for all the batch proving tasks to finish. Exit early on error.
+            while let Some(it) = batch_proving_futures.next().await {
+                it?;
+            }
+            Ok::<(), anyhow::Error>(())
```

Comment:
nit: looping over a stream
```suggestion
            let () = batch_proving_futures.try_collect().await?;
            anyhow::Ok(())
```
You might need a `map(|_| ())` if the `Ok` variant isn't already `()`